### PR TITLE
Feature sdk 571 resumable mpu new branch

### PR DESCRIPTION
**Pull request by twincitiesguy**

### CODE REVIEW

File: src/main/java/com/emc/object/s3/LargeFileUploader.java [hunk 1]
```diff
@@ -316,8 +517,29 @@ protected void configure() {
...
 
+        // if we are resuming an MPU, size is above MPU threshold, and we don't have an uploadId to resume,
+        // find an existing upload to resume
+        if (resumeContext != null && fullSize >= mpuThreshold) {
+            // find latest upload
+            if (resumeContext.getUploadId() == null) {
+                resumeContext.setUploadId(getLatestMultipartUploadId());
+            }
+            // list existing parts
+            if (resumeContext.getUploadId() != null && resumeContext.getPartsToSkip() == null) {
```
Comment (xiaoxin-ren) [inline src/main/java/com/emc/object/s3/LargeFileUploader.java:528]: One scenario needs to be considered: partsToSkip isn't checked if resumeIfInitiatedAfter is provided.

Comment (xiaoxin-ren) [inline src/main/java/com/emc/object/s3/LargeFileUploader.java:528]: One scenario needs to be considered: partsToSkip isn't checked if resumeIfInitiatedAfter is provided.

Comment (xiaoxin-ren) [inline src/main/java/com/emc/object/s3/LargeFileUploader.java:528]: One scenario needs to be considered: partsToSkip isn't checked if resumeIfInitiatedAfter is provided.

File: src/main/java/com/emc/object/s3/request/CopyObjectRequest.java [hunk 2]
```diff
@@ -54,8 +54,6 @@
     private AccessControlList acl;
     private CannedAcl cannedAcl;
 
-    private String copyMode;
```
Comment (twincitiesguy) [inline src/main/java/com/emc/object/s3/request/CopyObjectRequest.java:57]: Why is this change here?  Seems we are removing copyMode from CopyObjectRequest, but that should be in a separate branch/PR

Comment (xiaoxin-ren) [reply src/main/java/com/emc/object/s3/request/CopyObjectRequest.java:57]: It seems to be introduced by merging PR#75 which didn't pickup copyMode change in master.

File: src/main/java/com/emc/object/s3/LargeFileUploader.java [hunk 3]
```diff
@@ -138,84 +220,204 @@ public void upload() {
...
-            CompleteMultipartUploadResult result = s3Client.completeMultipartUpload(compRequest);
-            eTag = result.getETag();
+            eTag = completeMpu(uploadId, parts);
 
         } catch (Exception e) {
-
             // abort MP upload
             try {
-                s3Client.abortMultipartUpload(new AbortMultipartUploadRequest(bucket, key, uploadId));
+                abortMpu(uploadId);
```
Comment (xiaoxin-ren) [inline src/main/java/com/emc/object/s3/LargeFileUploader.java:412]: We probably shouldn't abort MPU if resume context is provided.

**Review by xiaoxin-ren — APPROVED** [review]

File: src/main/java/com/emc/object/s3/LargeFileUploader.java (hunk 3 above)
Comment (twincitiesguy) [reply src/main/java/com/emc/object/s3/LargeFileUploader.java]: this is handled by the new `abortMpuOnFailure` option

**Review by xiaoxin-ren — APPROVED** [review]

**Review by xiaoxin-ren — APPROVED** [review]
It looks good.
